### PR TITLE
[KYUUBI #304] Stop engine server gracefully when related zk node is deleted

### PR DESCRIPTION
**Pull request by turboFei**
turboFei [Closes #477](https://github.com/yaooqinn/kyuubi/pull/477)    Feature [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_
Stop engine server gracefully when related zk node is deleted

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala [hunk 1]
```diff
@@ -130,4 +130,45 @@ class ServiceDiscoverySuite extends KerberizedTestHelper {
...
+      assert(framework.checkExists().forPath("/abc") === null)
+      assert(framework.checkExists().forPath(znodeRoot) !== null)
+      val children = framework.getChildren.forPath(znodeRoot).asScala
+      assert(children.head ===
+        s"serviceUri=${server.connectionUrl};version=$KYUUBI_VERSION;sequence=0000000000")
+
+      children.foreach { child =>
+        framework.delete().forPath(s"""$znodeRoot/$child""")
+      }
+      Thread.sleep(5000)
```
Comment (yaooqinn) [inline kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala:165]: how about we use `eventually`?

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala [hunk 2]
```diff
@@ -130,4 +130,45 @@ class ServiceDiscoverySuite extends KerberizedTestHelper {
...
+      assert(serviceDiscovery.getServiceState === ServiceState.STOPPED)
+      assert(server.getServiceState === ServiceState.STOPPED)
```
Comment (yaooqinn) [inline kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala:167]: let also check the log？

**Review by yaooqinn — APPROVED** [review]
LGTM, only few minors

File: kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala (hunk 2 above)
Comment (turboFei) [reply kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/ServiceDiscoverySuite.scala]: ok

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala [hunk 3]
```diff
@@ -164,6 +173,21 @@ abstract class ServiceDiscovery private (
   // stop the server genteelly
   def stopGracefully(): Unit = {
     stop()
+    while (server.backendService != null &&
+      server.backendService.sessionManager.getOpenSessionCount > 0) {
+      Thread.sleep(1000 * 60)
```
Comment (turboFei) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala:178]: constant time waiting

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.2.0
